File: .sling/libs/sushi/eqela/jk-data/0.22.0/jk.base64.lua
jk = jk or {}

jk.base64 = jk.base64 or {}

jk.base64.Base64Decoder = {}
jk.base64.Base64Decoder.__index = jk.base64.Base64Decoder
_vm:set_metatable(jk.base64.Base64Decoder, {})

function jk.base64.Base64Decoder._create()
	local v = _vm:set_metatable({}, jk.base64.Base64Decoder)
	return v
end

function jk.base64.Base64Decoder:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.base64.Base64Decoder'
	self['_isType.jk.base64.Base64Decoder'] = true
end

function jk.base64.Base64Decoder:_construct0()
	jk.base64.Base64Decoder._init(self)
	return self
end

function jk.base64.Base64Decoder:fromUrlSafeFormat(str)
	local v = str
	if not (v ~= nil) then
		do return nil end
	end
	v = _g.jk.lang.String:replaceCharacter(v, 45, 43)
	v = _g.jk.lang.String:replaceCharacter(v, 95, 47)
	do
		local mod = _util:convert_to_integer(_g.jk.lang.String:getLength(v) % 4)
		if mod == 1 then
			v = _g.jk.lang.String:safeString(v) .. "==="
		elseif mod == 2 then
			v = _g.jk.lang.String:safeString(v) .. "=="
		elseif mod == 3 then
			v = _g.jk.lang.String:safeString(v) .. "="
		end
		do return v end
	end
end

function jk.base64.Base64Decoder:appendByte(buffer, byte)
	local nbyte = _g.jk.lang.Buffer:allocate(1)
	do _g.jk.lang.Buffer:setByte(nbyte, 0, byte) end
	do return _g.jk.lang.Buffer:append(buffer, nbyte, _g.jk.lang.Buffer:getSize(nbyte)) end
end

function jk.base64.Base64Decoder:fromLookupChar(ascii)
	local c = 0
	if ascii == 43 then
		c = 62
	elseif ascii == 47 then
		c = 63
	elseif ascii >= 48 and ascii <= 57 then
		c = ascii + 4
	elseif ascii >= 65 and ascii <= 90 then
		c = ascii - 65
	elseif ascii >= 97 and ascii <= 122 then
		c = ascii - 71
	end
	do return c end
end

function jk.base64.Base64Decoder:decode(text)
	if not (text ~= nil) then
		do return nil end
	end
	do
		local data = _util:allocate_buffer(0)
		local iter = _g.jk.lang.String:iterate(text)
		if iter ~= nil then
			local ch = 0
			while (function()
				ch = iter:getNextChar()
				do return ch end
			end)() > 0 do
				local c1 = 0
				local c2 = 0
				local c3 = 0
				local c4 = 0
				local byte1 = 0
				local byte2 = 0
				local byte3 = 0
				c1 = _g.jk.base64.Base64Decoder:fromLookupChar(ch)
				c2 = _g.jk.base64.Base64Decoder:fromLookupChar((function()
					ch = iter:getNextChar()
					do return ch end
				end)())
				byte1 = _vm:bitwise_left_shift(c1, 2) + _vm:bitwise_right_shift(c2, 4)
				data = _g.jk.base64.Base64Decoder:appendByte(data, byte1)
				ch = iter:getNextChar()
				if ch == 61 then
					do break end
				else
					c3 = _g.jk.base64.Base64Decoder:fromLookupChar(ch)
				end
				byte2 = _vm:bitwise_left_shift(_vm:bitwise_and(c2, 15), 4) + _vm:bitwise_right_shift(c3, 2)
				data = _g.jk.base64.Base64Decoder:appendByte(data, byte2)
				ch = iter:getNextChar()
				if ch == 61 then
					do break end
				else
					c4 = _g.jk.base64.Base64Decoder:fromLookupChar(ch)
				end
				byte3 = _vm:bitwise_left_shift(_vm:bitwise_and(c3, 3), 6) + c4
				data = _g.jk.base64.Base64Decoder:appendByte(data, byte3)
			end
		end
		do return data end
	end
end

jk.base64.Base64Encoder = {}
jk.base64.Base64Encoder.__index = jk.base64.Base64Encoder
_vm:set_metatable(jk.base64.Base64Encoder, {})

function jk.base64.Base64Encoder._create()
	local v = _vm:set_metatable({}, jk.base64.Base64Encoder)
	return v
end

function jk.base64.Base64Encoder:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.base64.Base64Encoder'
	self['_isType.jk.base64.Base64Encoder'] = true
end

function jk.base64.Base64Encoder:_construct0()
	jk.base64.Base64Encoder._init(self)
	return self
end

function jk.base64.Base64Encoder:toUrlSafeFormat(str)
	local v = str
	if not (v ~= nil) then
		do return nil end
	end
	v = _g.jk.lang.String:replaceString(v, "=", "")
	v = _g.jk.lang.String:replaceCharacter(v, 43, 45)
	v = _g.jk.lang.String:replaceCharacter(v, 47, 95)
	do return v end
end

function jk.base64.Base64Encoder:encodeString(str)
	if not (str ~= nil) then
		do return nil end
	end
	do return _g.jk.base64.Base64Encoder:encode(_g.jk.lang.String:toUTF8Buffer(str)) end
end

function jk.base64.Base64Encoder:encode(buffer)
	if not (buffer ~= nil) then
		do return nil end
	end
	do
		local length = _g.jk.lang.Buffer:getSize(buffer)
		local dst = _g.jk.lang.Buffer:allocate(_util:convert_to_integer((length + 2) / 3) * 4)
		local x = 0
		local dstIndex = 0
		local state = 0
		local old = 0
		do
			local srcIndex = 0
			while srcIndex < length do
				x = buffer[srcIndex + 1]
				do
					local v = (function() state = state + 1 return state end)()
					if v == 1 then
						dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(_g.jk.base64.Base64Encoder:toASCIIChar(_vm:bitwise_and(_vm:bitwise_right_shift(x, 2), 63)))
					elseif v == 2 then
						dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(_g.jk.base64.Base64Encoder:toASCIIChar(_vm:bitwise_or(_vm:bitwise_and(_vm:bitwise_left_shift(old, 4), 48), _vm:bitwise_and(_vm:bitwise_right_shift(x, 4), 15))))
					elseif v == 3 then
						dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(_g.jk.base64.Base64Encoder:toASCIIChar(_vm:bitwise_or(_vm:bitwise_and(_vm:bitwise_left_shift(old, 2), 60), _vm:bitwise_and(_vm:bitwise_right_shift(x, 6), 3))))
						dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(_g.jk.base64.Base64Encoder:toASCIIChar(_vm:bitwise_and(x, 63)))
						state = 0
					end
				end
				old = x
				do srcIndex = srcIndex + 1 end
			end
		end
		do
			local v2 = state
			if v2 == 1 then
				dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(_g.jk.base64.Base64Encoder:toASCIIChar(_vm:bitwise_and(_vm:bitwise_left_shift(old, 4), 48)))
				dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(61)
				dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(61)
			elseif v2 == 2 then
				dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(_g.jk.base64.Base64Encoder:toASCIIChar(_vm:bitwise_and(_vm:bitwise_left_shift(old, 2), 60)))
				dst[(function() local v = dstIndex dstIndex = dstIndex + 1 return v end)() + 1] = _util:convert_to_integer(61)
			end
		end
		do return _g.jk.lang.String:forBuffer(dst, "ASCII") end
	end
end

function jk.base64.Base64Encoder:toASCIIChar(lookup)
	local c = 0
	if lookup < 0 or lookup > 63 then
		do return c end
	end
	if lookup <= 25 then
		c = lookup + 65
	elseif lookup <= 51 then
		c = lookup + 71
	elseif lookup <= 61 then
		c = lookup - 4
	elseif lookup == 62 then
		c = _util:convert_to_integer(43)
	elseif lookup == 63 then
		c = _util:convert_to_integer(47)
	end
	do return c end
end

File: .sling/libs/web5/eqela/jk-data/0.32.0/jk.base64.js
let JkBase64Base64Decoder = function() {};

JkBase64Base64Decoder.prototype._t = { "JkBase64Base64Decoder" : true };
JkBase64Base64Decoder.prototype._tobj = JkBase64Base64Decoder;

JkBase64Base64Decoder.NEW = function() {
	var v = new JkBase64Base64Decoder;
	return v.CTOR_JkBase64Base64Decoder();
};

JkBase64Base64Decoder.prototype.CTOR_JkBase64Base64Decoder = function() {
	return this;
};

JkBase64Base64Decoder.fromUrlSafeFormat = function(str) {
	var v = str;
	if(!(v != null)) {
		return null;
	}
	v = JkLangString.replaceCharacter(v, '-'.charCodeAt(), '+'.charCodeAt());
	v = JkLangString.replaceCharacter(v, '_'.charCodeAt(), '/'.charCodeAt());
	var mod = JkLangString.getLength(v) % 4;
	if(mod == 1) {
		v = (JkLangString.safeString(v)) + ("===");
	}
	else if(mod == 2) {
		v = (JkLangString.safeString(v)) + ("==");
	}
	else if(mod == 3) {
		v = (JkLangString.safeString(v)) + ("=");
	}
	return v;
};

JkBase64Base64Decoder.appendByte = function(buffer, _byte) {
	var nbyte = JkLangBuffer.allocate(1);
	JkLangBuffer.setByte(nbyte, 0, _byte);
	return JkLangBuffer.append(buffer, nbyte, (JkLangBuffer.getSize(nbyte)));
};

JkBase64Base64Decoder.fromLookupChar = function(ascii) {
	var c = 0;
	if(ascii == 43) {
		c = 62;
	}
	else if(ascii == 47) {
		c = 63;
	}
	else if(ascii >= 48 && ascii <= 57) {
		c = ascii + ~(~4);
	}
	else if(ascii >= 65 && ascii <= 90) {
		c = ascii - ~(~65);
	}
	else if(ascii >= 97 && ascii <= 122) {
		c = ascii - ~(~71);
	}
	return c;
};

JkBase64Base64Decoder.decode = function(text) {
	if(!(text != null)) {
		return null;
	}
	try {
		var binaryString = window.atob(text);
		var len = binaryString.length;
		var buffer = new ArrayBuffer(len);
		var arr = new Uint8Array(buffer);
		for(var i = 0; i < len; i++) {
			arr[i] = binaryString.charCodeAt(i);
		}
		if(buffer != null) {
			return buffer;
		}
	} catch(e) {
		console.log("Exception: " + e.message);
		return null;
	}
	;
	return null;
};

JkBase64Base64Decoder.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkBase64Base64Decoder"] === true;
};

let JkBase64Base64Encoder = function() {};

JkBase64Base64Encoder.prototype._t = { "JkBase64Base64Encoder" : true };
JkBase64Base64Encoder.prototype._tobj = JkBase64Base64Encoder;

JkBase64Base64Encoder.NEW = function() {
	var v = new JkBase64Base64Encoder;
	return v.CTOR_JkBase64Base64Encoder();
};

JkBase64Base64Encoder.prototype.CTOR_JkBase64Base64Encoder = function() {
	return this;
};

JkBase64Base64Encoder.toUrlSafeFormat = function(str) {
	var v = str;
	if(!(v != null)) {
		return null;
	}
	v = JkLangString.replaceString(v, "=", "");
	v = JkLangString.replaceCharacter(v, '+'.charCodeAt(), '-'.charCodeAt());
	v = JkLangString.replaceCharacter(v, '/'.charCodeAt(), '_'.charCodeAt());
	return v;
};

JkBase64Base64Encoder.encodeString = function(str) {
	if(!(str != null)) {
		return null;
	}
	return JkBase64Base64Encoder.encode((JkLangString.toUTF8Buffer(str)));
};

JkBase64Base64Encoder.encode = function(buffer) {
	if(!(buffer != null)) {
		return null;
	}
	var base64 = null;
	var bin = '';
	var bytes = new Uint8Array( buffer );
	var len = bytes.byteLength;
	for (var i = 0; i < len; i++) {
		bin += String.fromCharCode( bytes[i] );
	}
	base64 = window.btoa(bin);
	;
	return base64;
};

JkBase64Base64Encoder.toASCIIChar = function(lookup) {
	var c = 0;
	if(lookup < 0 || lookup > 63) {
		return c;
	}
	if(lookup <= 25) {
		c = lookup + 65;
	}
	else if(lookup <= 51) {
		c = lookup + 71;
	}
	else if(lookup <= 61) {
		c = lookup - 4;
	}
	else if(lookup == 62) {
		c = ~(~'+'.charCodeAt());
	}
	else if(lookup == 63) {
		c = ~(~'/'.charCodeAt());
	}
	return c;
};

JkBase64Base64Encoder.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkBase64Base64Encoder"] === true;
};
